src/vm-vcpu-ref/src/aarch64/regs/icc.rs

```diff
@@ -0,0 +1,187 @@
+            | (((crm as u64) << KVM_REG_ARM64_SYSREG_CRM_SHIFT)
+                & KVM_REG_ARM64_SYSREG_CRM_MASK as u64)
+            | (((op2 as u64) << KVM_REG_ARM64_SYSREG_OP2_SHIFT)
+                & KVM_REG_ARM64_SYSREG_OP2_MASK as u64);
+
+        SimpleReg { offset, size: 8 }
+    }
+
+    const fn sys_icc_ap0rn_el1(n: u64) -> SimpleReg {
+        Self::gic_sys_reg(3, 0, 12, 8, 4 | n)
+    }
+
+    const fn sys_icc_ap1rn_el1(n: u64) -> SimpleReg {
+        Self::gic_sys_reg(3, 0, 12, 9, n)
+    }
+}
+
+/// Get vCPU GIC system registers.
+pub fn icc_regs(fd: &DeviceFd, mpidr: u64) -> Result<GicSysRegsState> {
+    let main_icc_regs = get_regs_data(
+        fd,
+        MAIN_GIC_ICC_REGS.iter(),
+        KVM_DEV_ARM_VGIC_GRP_CPU_SYSREGS,
+        mpidr,
+        KVM_DEV_ARM_VGIC_V3_MPIDR_MASK as u64,
+    )?;
+
+    let num_priority_bits = num_priority_bits(fd, mpidr)?;
+
+    let mut ap_icc_regs = Vec::with_capacity(AP_GIC_ICC_REGS.len());
+    for reg in AP_GIC_ICC_REGS {
+        if is_ap_reg_available(reg, num_priority_bits) {
+            ap_icc_regs.push(Some(get_reg_data(
+                fd,
+                reg,
+                KVM_DEV_ARM_VGIC_GRP_CPU_SYSREGS,
+                mpidr,
+                KVM_DEV_ARM_VGIC_V3_MPIDR_MASK as u64,
+            )?));
+        } else {
+            ap_icc_regs.push(None);
+        }
+    }
+
+    Ok(GicSysRegsState {
+        main_icc_regs,
+        ap_icc_regs,
+    })
+}
+
+/// Set vCPU GIC system registers.
+pub fn set_icc_regs(fd: &DeviceFd, state: &GicSysRegsState, mpidr: u64) -> Result<()> {
+    set_regs_data(
+        fd,
+        MAIN_GIC_ICC_REGS.iter(),
+        KVM_DEV_ARM_VGIC_GRP_CPU_SYSREGS,
+        &state.main_icc_regs,
+        mpidr,
+        KVM_DEV_ARM_VGIC_V3_MPIDR_MASK as u64,
+    )?;
+
+    let num_priority_bits = num_priority_bits(fd, mpidr)?;
+
+    for (reg, maybe_reg_data) in AP_GIC_ICC_REGS.iter().zip(&state.ap_icc_regs) {
+        if is_ap_reg_available(reg, num_priority_bits) != maybe_reg_data.is_some() {
+            return Err(Error::InvalidGicSysRegState);
+        }
+
+        if let Some(reg_data) = maybe_reg_data {
+            set_reg_data(
+                fd,
+                reg,
+                KVM_DEV_ARM_VGIC_GRP_CPU_SYSREGS,
+                reg_data,
+                mpidr,
+                KVM_DEV_ARM_VGIC_V3_MPIDR_MASK as u64,
+            )?;
+        }
+    }
+
+    Ok(())
+}
+
+fn num_priority_bits(fd: &DeviceFd, mpidr: u64) -> Result<u64> {
+    let reg_val: u64 = get_reg_data(
+        fd,
+        &SYS_ICC_CTLR_EL1,
+        KVM_DEV_ARM_VGIC_GRP_CPU_SYSREGS,
+        mpidr,
+        KVM_DEV_ARM_VGIC_V3_MPIDR_MASK as u64,
+    )?
+    .chunks[0];
+
+    Ok(((reg_val & ICC_CTLR_EL1_PRIBITS_MASK) >> ICC_CTLR_EL1_PRIBITS_SHIFT) + 1)
+}
+
+fn is_ap_reg_available(reg: &SimpleReg, num_priority_bits: u64) -> bool {
+    // As per ARMv8 documentation:
+    // https://static.docs.arm.com/ihi0069/c/IHI0069C_gic_architecture_specification.pdf
```

Comment:
This link does not work for me. Do you find the documentation in some other place as well?